### PR TITLE
Warn user if Parameter parses something other than a string.

### PR DESCRIPTION
**Pull request by dlstadther**
See within thread for Issue #1607

### CODE REVIEW

**Comment by Tarrasch** [comment]
But this is going to give so many warnings to so many users!!!



... and I like it :sunglasses:!

File: luigi/parameter.py [hunk 1]
```diff
@@ -219,6 +219,8 @@ def parse(self, x):
         :param str x: the value to parse.
         :return: the parsed value.
         """
+        if not isinstance(x, basestring) or not isinstance(x, str):
```
Comment (dlstadther) [inline luigi/parameter.py]: My mac is not playing nice installing python 3.5 and setting up a venv, but does anyone know if this line will falsely evaluate that a string (unicode or ascii) is not an instance of basestring for python 3+?

This does seem to work correctly for python 2.7.

Comment (dlstadther) [inline luigi/parameter.py]: Or apparently all the python 3 tests will fail because they don't know what `basestring` is.

How should one go about this condition? I need `if not isinstance(x, basestring)` for python2 and `if not isinstance(x, str)` for python3.

Comment (dlstadther) [inline luigi/parameter.py]: I can only think of the following:

```python
def parse(self, x):
     try:  # try python2 syntax
          if not isinstance(x, basestring):
               warnings.warn(...)
      except:  # if py2 syntax fails, try py3
          if not isinstance(x, str):
               warnings.warn(...)
      return x
```

But this repeats `warnings.warn(...)` :disappointed:

**Comment by dlstadther** [comment]
Opted for the duplicate warning message since i'm trying to get comments and our work schedules appear to complete opposites (yay timezones).

I'm very open to better suggestions.

**Comment by Tarrasch** [comment]
Have you checked in six.py? In particular `six.string_types` looks relevant.

**Comment by Tarrasch** [comment]
Once you've checked the six.py module, let me know when you think this is merge-worthy.

**Comment by dlstadther** [comment]
I have not. But now that I know to look there, I will and will notify you when complete. Thanks

**Comment by dlstadther** [comment]
I don't want to change how python 2 handles `str` or `unicode` nor python 3 `str`. So unless someone else believes i'm mistaken, I don't believe any change is necessary within six.py.

File: luigi/parameter.py [hunk 2]
```diff
@@ -219,7 +219,14 @@ def parse(self, x):
...
         :return: the parsed value.
         """
-        return x  # default impl
+        try:
+            if not isinstance(x, basestring):   # python 2
+                warnings.warn("Parameter %r is not of type string." % x)
+        except:
+            if not isinstance(x, str):          # python 3
+                warnings.warn("Parameter %r is not of type string." % x)
+        finally:
```
Comment (dlstadther) [inline luigi/parameter.py]: However, i'm open to suggestion on how to clean this up and not have the `warnings.warn(...)` message written twice.

The only other thing i can think of to only write the message once is:
```python
warn = False
try:
    if not isinstance(x, basestring):
        warn = True
except:
    if not isinstance(x, str):
        warn = True
finally:
    if warn:
        warnings.warn("Parameter %r is not of type string." % x)
    return x
```

**Comment by erikbern** [comment]
https://pythonhosted.org/six/#six.string_types

you can just do `if isinstance(value, six.string_Types)`

**Comment by dlstadther** [comment]
Oh ok! I didn't realize that's how six.py was supposed to be used. Thanks @erikbern !

**Comment by Tarrasch** [comment]
I was about to merge this, until it I just realized that I don't understand this patch. Can you reiterate in this PR why you are doing this? I thought we wanted to check that nobody passes a non-string to Parameter.

**Comment by dlstadther** [comment]
**tl;dr**: `Parameter` should really be named `StrParameter` since it de/serializes strings; however, when using normal requirements this can be disregarded without unintended consequences, but misuse of `Parameter` within dynamic requirements will result in data treated as strings.

Through using dynamic requires, I found that anything passed as a `Parameter` was treated as a string (which is why I added PR #1614 ). 

Parameters take one of two paths within luigi execution: 1. dynamic requires 2. normal requires.

1. @davemt found that

> dynamic dependencies have `load_task` called on them before they are added to worker's task cache whereas regular `requires()` do not. This requires that a parameter value can be properly serialized/deserialized.

So since the `Parameter` is first serialized, it is cast as a string and then parsed (as a string). Thus no warning message.

2. However, during normal requires, there is no guaranteed serialization/deserialization. So you can (as I have historically) incorrectly pass any type of structure through as a `Parameter` (i.e. Lists, Tuples, etc.). 
So since `Parameter` assumes strings (as seen by its `serialize()` method), @erikbern suggested

> It might be good to do a warnings.warn in case serialize returns a non-string

**Comment by dlstadther** [comment]
@Tarrasch , any comments or questions?

**Comment by Tarrasch** [comment]
I stared at this for a while and I still don't understand this patch.

When could `parse()` ever be called with a non-string? That would only happen on fatal framework (luigi) errors. No?

Why not just put the check in `serialize` where it intuitively makes sense? That would occur on actual user errors.

**Comment by dlstadther** [comment]
So @Tarrasch is correct, the string check within`parse` won't warn the user. However, we cannot just simply move that string check into `serialize`.

The problem with including the warn within `serialize` is that other Parameters inherit (and don't override) that method. e.g. `FloatParameter`, `IntParameter`, and `BoolParmeter`.

If we add the string check within `Parameter.serialize`, we will need to override the `serialize` method for int, float, and bool.

**Comment by Tarrasch** [comment]
> If we add the string check within Parameter.serialize, we will need to override the serialize method for int, float, and bool.

This is not correct I think. There is an easier way, just check if the current class is Parameter or not.

```python
if not isinstance(x, six.string_types) and self.class == Parameter:
```

**Comment by dlstadther** [comment]
Learn something new everyday. Thanks! I'll add that tomorrow morning.

**Comment by Tarrasch** [comment]
Thanks. Did you test to see if this works as expected? If so let's merge :)

**Comment by dlstadther** [comment]
Yes, I tested using the following code:

```python
import datetime

import luigi
from luigi import mock

__author__ = "Dillon Stadther"
__date__ = '2016-03-24'


class TestTask(luigi.Task):
    date = luigi.DateParameter(default=datetime.date(2016, 03, 24))

    def requires(self):
        yield TestClass(
            date=self.date,                # prints "2016-03-24" and "2016-03-17"
            string_in='Hello World',       # prints "Hello World" and "H"
            str1_in=['Bye'],
            list_in=['Hello', 'World'],    # prints "('Hello', 'World')" and "Hello"
            tuple_in=('foo', 'bar'),       # prints "('foo', 'bar')" and "foo"
            int_in=10                      # prints "10" and "20"
        )

    def output(self):
        return mock.MockTarget('test_requires')

    def run(self):
        yield TestClass(
            date=self.date,                 # prints "2016-03-24" and "2016-03-17"
            string_in='Hello World',        # prints "Hello World" and "H"
            str1_in=['Bye', 'Doge'],
            list_in=['Hello', 'World'],     # prints "('Hello', 'World')" and "("
            tuple_in=('foo', 'bar'),        # prints "('foo', 'bar')" and "("
            int_in=10                       # prints "10" and "20"
        )
        self.output().open('w').close()


class TestClass(luigi.Task):
    date = luigi.DateParameter()

    string_in = luigi.Parameter(default='')
    str1_in = luigi.Parameter(default='')
    list_in = luigi.ListParameter(default=[])
    tuple_in = luigi.TupleParameter(default=())
    int_in = luigi.IntParameter(default=0)

    def output(self):
        return mock.MockTarget('test_out')

    def run(self):
        print(self.date)
        print(self.date + datetime.timedelta(days=-7))  # should print the date 7 days ago
        print(self.string_in)
        print(self.string_in[0])        # should print first character
        print(self.str1_in)
        print(self.str1_in[0])  # should print first character
        print(self.list_in)
        print(self.list_in[0])          # should print first element
        print(self.tuple_in)
        print(self.tuple_in[0])         # should print first element
        print(self.int_in)
        print(self.int_in * 2)          # should print double the int

        self.output().open('w').close()


if __name__ == "__main__":
    luigi.run(['TestTask', '--workers', '1'])
```

When `str1_in` is serialized (for both normal requires and dynamic requires), the following warning messages are presented:

```
/Users/dstadther/venv-luigi/lib/python2.7/site-packages/luigi-2.1.1-py2.7.egg/luigi/parameter.py:240: UserWarning: Parameter ('Bye',) is not of type string.
  warnings.warn("Parameter {0} is not of type string.".format(str(x)))
/Users/dstadther/venv-luigi/lib/python2.7/site-packages/luigi-2.1.1-py2.7.egg/luigi/parameter.py:240: UserWarning: Parameter ('Bye', 'Doge') is not of type string.
  warnings.warn("Parameter {0} is not of type string.".format(str(x)))
```

**Comment by erikbern** [comment]
bonus points if you want to add a unit test for this. but i'm fine mergin as-is

**Comment by dlstadther** [comment]
@erikbern Bonus points have been attempted

**Comment by erikbern** [comment]
fantastic!

File: test/parameter_test.py [hunk 3]
```diff
@@ -70,6 +71,34 @@ def run(self):
...
+    def requires(self):
+        # normal requires
+        yield BlankTask(str_param=self.str_param, non_str_param=self.non_str_param)
+
+    def output(self):
+        return MockTarget("dumb_output")
+
+    def run(self):
+        # dynamic requires
+        yield DumbTask2(str_param=self.str_param, non_str_param=self.non_str_param)
```
Comment (erikbern) [inline test/parameter_test.py]: is DumbTask2 defined? Flake8 is complaining about it

File: test/parameter_test.py [hunk 4]
```diff
@@ -285,6 +314,14 @@ def test_tuple_serialize_parse(self):
         b_tuple = ((1, 2), (3, 4))
         self.assertEqual(b_tuple, a.parse(a.serialize(b_tuple)))
 
+    def test_requires_warn_serialize(self):
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            NonStrParamTask()
```
Comment (erikbern) [inline test/parameter_test.py]: what's this statement useful for?

File: test/parameter_test.py [hunk 5]
```diff
@@ -285,6 +314,14 @@ def test_tuple_serialize_parse(self):
...
+            assert len(w) == 1
+            assert issubclass(w[-1].category, Warning)
+            assert "is not of type string." in str(w[-1].message)
```
Comment (erikbern) [inline test/parameter_test.py]: you should use unittest assertions here – self.assertEquals etc

**Comment by erikbern** [comment]
for some reason the unit tests are failing in 2.7 and 3.3 – not sure if it's because we get 0 warnings or 2+

File: test/parameter_test.py (hunk 3 above)
Comment (dlstadther) [inline test/parameter_test.py]: I changed the name of the class and forgot to update all calls to it.

**Comment by dlstadther** [comment]
This is weird...

I have corrected your in-line comments and reran locally. I get a :+1: for `tox -e py27-nonhdfs test/parameter_test.py`, but a :-1: for `tox -e py27-nonhdfs`. The difference in failure is with asserting that there is 1 warning.

Any idea why this may be occurring?

**Comment by erikbern** [comment]
Not sure. If you can't figure it out, no problem. The test isn't super important right now, we can just remove the last commit

**Comment by dlstadther** [comment]
I'm going to mess with this for a bit and i'll report back by tomorrow.

**Comment by dlstadther** [comment]
I'm trying to understand the process of execution for `tox -e py27-nonhdfs`.

So tox is install on my local system (specifically within my venv). Tox reads the tox.ini file and creates a virtual testing environment based on the .ini file. It then runs the commands that are listed - which appears to be (optionally setup hadoop environment), print python version, measure the coverage of tests within the project by executing `test/runtests.py, combine coverage data, and then doing something with codecov.

[`test/runtests.py`](https://github.com/spotify/luigi/blob/master/test/runtests.py) appears to...
1. catch warnings using the `default` flag (according to [py `warnings` docs](https://docs.python.org/2/library/warnings.html#the-warnings-filter)...`default` will "print the first occurrence of matching warnings for each location where the warning is issued")  
2. ignore the printing of `UserWarning`  
3. execute `nose.main()` (...which does what exactly???) (If someone could explain this, that'd be much appreciated!)  

Somehow this executes all files within `test/*`, but I'm unsure how.

I'm also unsure how specifying a filename after the environment (`tox -e py27-nonhdfs test/parameter_test.py`) changes this execution.

My concern is that the environment for all tests is somehow different from that for a single test. If possible, I'd feel most comfortable getting a successful test prior to merging (in the case that something isn't quite right and spams everyone with false positive warning messages).

**Comment by erikbern** [comment]
i'm not sure if there's an easy way to run a single test using tox. unfortunately i generally have pretty bad experience trying to use tox.

i'm fine if you just want to remove the last commit from this PR. we don't have to unit test warnings :)

**Comment by dlstadther** [comment]
Yeah, tox documentation isn't very clear (and yet is at least semi-active - latest release < 6 months ago).

I'll remove the tests.

I just ran some other jobs using this PR and am getting a warning saying that `None is not of type string` although i'm not passing None as any Parameter. So i'm not comfortable merging this yet.

**Comment by Tarrasch** [comment]
@dlstadther, Oh I guess that's because there's part of luigi that itself has "None" has default value. You should fix those cases too before we merge this. Otherwise users will get confused by unavoidable warnings.

**Comment by erikbern** [comment]
ok that makes sense. yeah since warnings.warn only triggers the first time, it's probably because it was already run somewhere else. good catch, @Tarrasch

**Comment by dlstadther** [comment]
@Tarrasch I was getting mildly suspicious of such. Do you know off the top of your head where this "part of luigi" might be location? As time allows today, I'll do some searching.

**Comment by erikbern** [comment]
it should be easy to find if you just replace the warning with an exception and run the tests :)

**Comment by dlstadther** [comment]
The culprit: a param named `scheduler_url` with a value of `None`. It is included in the list of other Luigi system Parameters which are returned by `dir(cls)` in `Task.py` `get_params(cls)` [L210](https://github.com/spotify/luigi/blob/master/luigi/task.py#L210).

``` python
@classmethod
    def get_params(cls):
        """
        Returns all of the Parameters for this Task.
        """
        # We want to do this here and not at class instantiation, or else there is no room to extend classes dynamically
        params = []
        for param_name in dir(cls):
            param_obj = getattr(cls, param_name)
            if not isinstance(param_obj, Parameter):
                continue

            params.append((param_name, param_obj))

        # The order the parameters are created matters. See Parameter class
        params.sort(key=lambda t: t[1]._counter)
        return params
```

Based on Luigi docs, `scheduler_url` should default to `http://localhost:8082` (unless I'm understanding this incorrectly), but execution shows that `scheduler_url` (at this point in execution) is set to `None`.

**Comment by erikbern** [comment]
there must be a reason we set them to None. not sure

**Comment by dlstadther** [comment]
If it's valid to have optional Luigi system Parameters as `None`, but not for User-Defined Parameters to be `None`, Can there just be an additional check that the `Parameter` is not a string and not `None` before it warns the user?

**Comment by dlstadther** [comment]
`scheduler_url` looks to be set to `None` within interface.py so that it can be dynamically created via `scheduler_host` and `scheduler_port`.

**Comment by dlstadther** [comment]
Scratch that last statement.

The same thing happens with `module`, but that appears to be involved with CLI parsing (or at least that's what `always_in_help=True` makes be believe).

**Comment by dlstadther** [comment]
```python
 if not isinstance(x, six.string_types) and self.__class__ == Parameter and x is not None:
      warnings.warn("Parameter {0} is not of type string.".format(str(x)))
```

Does not warn against these Luigi system parameters, but also does not warn against user-defined Parameters that are defined as `None`. Are there any problems with this solution?

**Comment by Tarrasch** [comment]
@dlstadther, can you just change those `default=None` to `default=''` then? If there is any `param is None` checks just change them to `!= ''`. Should be doable right?

**Comment by dlstadther** [comment]
I can definitely do that for `scheduler_url`. I'm unsure what the significance of `module=None`. So I can, but there may be unknown side-effects if the CLI parser's help (or whatever `always_in_help` insinuates) expects a potential `None` value.

**Comment by Tarrasch** [comment]
I don't think there's any "gotcha" for  the `module` parameter. Don't mind `always_in_help`. There's both tests and human review to catch any mistakes. :)

**Comment by dlstadther** [comment]
Tests (luigi unit tests and my jobs) ran successfully and as expected locally. Hopefully the same will reign true for Travis-CI.

**Comment by erikbern** [comment]
looks like your unit test disappeared?

**Comment by dlstadther** [comment]
yeah, you told me to remove them since I was receiving different results depending on whether I ran just the `parameter_test.py` or the entire nonhdfs suite of tests.

**Comment by dlstadther** [comment]
I can spend more time on testing if necessary

**Comment by erikbern** [comment]
i thought you fixed the issue causing the unit tests to break? i think if you add back the unit tests they will pass right?

anyway, it's fine, let's merge this!

**Comment by dlstadther** [comment]
I'll the test one more shot. Even if it's post-merge.

**Comment by dlstadther** [comment]
For the life of me, I still cannot figure out why the below unit test passes for `tox -e py27-nonhdfs test/parameter_test.py`, but fails when the entire suite of nonhdfs tests are executed.

```python
    def test_requires_warn_serialize(self):
        class NonStrParamTask(luigi.Task):
            str_param = luigi.Parameter(default='Hello World')
            non_str_param = luigi.Parameter(default=1)

            def output(self):
                return MockTarget("dumb_output")

            def run(self):
                self.output().open('w').close()

        import warnings
        with warnings.catch_warnings(record=True) as w:
            warnings.simplefilter("always")
            NonStrParamTask(str_param="Hello World", non_str_param=1)
            self.assertEqual(len(w), 1)
            self.assertTrue(issubclass(w[-1].category, UserWarning))
            self.assertTrue("is not of type string." in str(w[-1].message))
```
```
FAIL: test_requires_warn_serialize (parameter_test.ParameterTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/dstadther/venv-l-test/luigi/test/parameter_test.py", line 304, in test_requires_warn_serialize
    self.assertEqual(len(w), 1)
AssertionError: 0 != 1
```

**Comment by erikbern** [comment]
that's really weird. maybe something else triggers the warning earlier?

**Comment by ulzha** [comment]
Do we not want people to use `None`, as a parameter default value, in general? I've treated `default=None` as a fitting way to specify that a parameter is optional (can have a value or have no value).

**Comment by Tarrasch** [comment]
@ulzha I found it a bit unmathematical and it causes bugs all over. In particular string value `None` and `'None'` both serializes to `'None'` and that only de-serializes to `'None'`. The `inverse_of(serialize) == parse` invariant is broken.

In most real world cases, there's probably a suitable default value that's a valid set-member of the type of the parameter (typically values in Python corresponding to `None` like `''`, `0`, `[]` etc).
